src/Engine/Data/Texture.hpp

```diff
@@ -16,230 +17,288 @@ class Texture;
 };
 
-/** Represent a Texture of the engine
+/// ImageParameters stores the infomation needed to upload a image to the GPU as a texture.
+/// The actual data is stored as a std::variant, so this image parameter can storke either a single
+/// image or a cube map as an array of 6 images. The stored shared_ptr is void, and must correspond
+/// to the format described by the other ImageParameters data member.
+/// Its perfectly fine to have nullptr texels to represent GPU only texture.
+struct ImageParameters {
+    /// Types for texels variant
+    using ImageType   = std::shared_ptr<void>;
+    using CubeMapType = std::array<std::shared_ptr<void>, 6>;
+
+    /// check which type is held by texels
+    template <typename TexelType>
+    bool isTexelOfType() const {
+        return std::holds_alternative<TexelType>( texels );
+    }
+
+    ImageType getImage() const {
+        CORE_ASSERT( isTexelOfType<ImageType>(), "texture variant is not ImageType" );
+        return std::get<ImageType>( texels );
+    }
+    void setImage( ImageType image ) { texels = image; }
+
+    /// get texels raw pointer
+    /// throws std::bad_variant_access if !isTexelOfType<ImageType>()
+    /// see std::get
+    const void* getTexels() const { return getImage().get(); }
+
+    /// get cube map array of shared ptr
+    /// throws std::bad_variant_access if !isTexelOfType<CubeMapType>()
+    /// see std::get
+    const CubeMapType getCubeMap() const {
+        CORE_ASSERT( isTexelOfType<CubeMapType>(), "texture variant is not CubeMapType" );
+        return std::get<CubeMapType>( texels );
+    }
+    void setCubeMap( CubeMapType cubeMap ) { texels = cubeMap; }
+
+    GLenum target { GL_TEXTURE_2D }; //< OpenGL target
+    size_t width { 1 };              //< width of the texture (s dimension)
+    size_t height { 1 };             //< height of the texture (t dimension)
+    size_t depth { 1 };              //< depth of the texture (r dimension)
+    GLenum format { GL_RGB };        //< Format of the external data
+    /// OpenGL internal format (WARNING, for Integer textures, must be GL_XXX_INTEGER)
+    GLenum internalFormat { GL_RGB };
+    GLenum type { GL_UNSIGNED_BYTE }; //< Type of the components in external data
+    /// set to true when linearize texture rgb component. If true, linearize has no effect.
+    bool isLinear { false };
+    std::variant<ImageType, CubeMapType> texels { nullptr }; //< texels OR cubeMap, shared ownership
+};
+
+/** @brief Describes the sampler and image of a texture.
+ *
+ * These parameters describe the image data of the texture :
+ *    - ImageParameters: target, width, height, depth, format, internalFormat, type and texels for
+ * describing image data
+ *    - SampleParameters: wrapS, wrapT, wrapP, minFilter and magFilter for describing the sampler of
+ * the texture.
+ *
+ * MipMap representation of the texture is automatically generated as soon as the minFilter
+ * parameter is something else than GL_LINEAR or GL_NEAREST
```

Comment:
```suggestion
 * parameter is something else than GL_LINEAR or GL_NEAREST.
 * @todo allow to use another mipmap building function than glGenerateMipmap
```
Mipmap could be built in another way then just averaging on pixels as done by glGenerateMipmap. Delay this to a future use case (I've some) but remind here this future evolution.